{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "205676c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from os.path import join\n",
    "import pandas as pd\n",
    "import xlsxwriter\n",
    "import openpyxl\n",
    "import os\n",
    "import numpy as np\n",
    "import itertools\n",
    "import itertools as it\n",
    "from pandas import Series, ExcelWriter\n",
    "import csv\n",
    "from skbio.stats.composition import clr\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acb70d15",
   "metadata": {},
   "source": [
    "#### Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "02950043",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_path = 'C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/'\n",
    "# top_path = 'C:/Users/morts/Documents/TAU/RF_Project/Data/'\n",
    "\n",
    "db_path1 = top_path + 'Ravel_et_al.2013/Hiseq_sequencing/DADA2_results/'\n",
    "db_path2 = top_path + 'Ravel_et_al.2013/454_Sequencing/DADA2_results/DADA_output_files/'\n",
    "db_path3 = top_path + 'Carter_et_al.2022/DADA2_results/'\n",
    "db_path4 = top_path + 'Ravel_et_al.2011/DADA2_results/'\n",
    "db_path5 = top_path + 'Serrano_et_al.2019/noDuplicatedID/DADA2_results/'\n",
    "db_path6 = top_path + 'Srinivasan2012/DADA2_results/'\n",
    "# db_path3 = top_path + 'Ceccarani2019/DADA2_results/'\n",
    "\n",
    "\n",
    "asv_abun_file_name = 'abundance_ASV.csv'\n",
    "taxa_file_name = 'taxa_ASV.csv'\n",
    "\n",
    "speciate_filename1 = 'Ravel2013_Hiseq_2step_single_V3V4_MC_order7.txt'\n",
    "speciate_filename2 = 'Ravel2013_V1V3_454_all_MC_order7_rev.txt'\n",
    "speciate_filename3 = 'Carter2022_V3V4_MC_order7.txt'\n",
    "speciate_filename4 = 'Ravel2011_V1V3_MC_order7.txt'\n",
    "speciate_filename5 = 'Serrano2019_MC_order7.txt'\n",
    "speciate_filename6 = 'Srinivasan2012_V3V4_MC_order7.txt'\n",
    "# speciate_filename3 = 'Ceccarani2019_V3V4_MC_order7.txt'\n",
    "\n",
    "taxa_names_lst = ['k','p','c','o','f','g','s']\n",
    "names_lst = ['Lactobacillus', 'Gardnerella', 'Prevotella', 'Atopobium', 'Shuttleworthia', 'Sneathia', 'BVAB', 'Unknown']\n",
    "date = '16052023'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f5820f",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a01f7aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def order_tax(df):\n",
    "    new_df = df.copy()\n",
    "    new_df = df.fillna(0)\n",
    "    new_df.columns = taxa_names_lst\n",
    "    new_df['cond_spec'] = np.nan\n",
    "    \n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "841ebba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cond(row):\n",
    "    if type(row['s']) == str:\n",
    "        row['cond_spec'] = row['g'] + '_' + row['s']\n",
    "    \n",
    "    else:\n",
    "        for i in range(len(taxa_names_lst)):\n",
    "            curr_level = taxa_names_lst[i]\n",
    "            if row[curr_level] == 0:\n",
    "                if curr_level == 'k':\n",
    "                    row['cond_spec'] = np.nan\n",
    "                else:\n",
    "                    upper_level = taxa_names_lst[i - 1]\n",
    "                    row['cond_spec'] = upper_level + '_' + row[upper_level]\n",
    "                break\n",
    "    \n",
    "    return row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dbc95d00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def order_speciate(df):\n",
    "    df.columns = ['ASV', 'speciate_spec', 'posterior_probability', 'decisions_num']\n",
    "    df.drop(['posterior_probability', 'decisions_num'], axis = 1, inplace = True)\n",
    "    df.set_index('ASV', inplace = True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c68ebe4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_final_name(row):\n",
    "    if any(name in row['speciate_spec'] for name in names_lst) or 'BVAB' in row['speciate_spec']:\n",
    "        return row['speciate_spec']\n",
    "    else:\n",
    "        if row['g'] is not 'Unknown' and '00' not in row['g']:\n",
    "            return 'g_' + row['g']\n",
    "        elif row['f'] != 0:\n",
    "            return 'f_' + row['f']\n",
    "        elif row['s'] == 0: \n",
    "            return row['cond_spec']\n",
    "        else:\n",
    "            return 'Unknown'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "84b7be19",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_new_name(name):\n",
    "    if name.split('_')[0] == 'f':\n",
    "        return name\n",
    "    elif 'Cluster' in name:\n",
    "        return name\n",
    "    elif len(name.split('_')[0]) == 1 and name.split('_')[0] != 'f':\n",
    "        return '_'.join(name.split('_')[:2])\n",
    "    elif any(map(str.isdigit, name.split('_')[1])):\n",
    "        return '_'.join([name.split('_')[i] for i in [0, 2]])\n",
    "    else: \n",
    "        return '_'.join(name.split('_')[-2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d350117b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spec_df(path, spec_path):\n",
    "    ## DADA tax df\n",
    "    tax_df_orig = pd.read_csv(path + taxa_file_name, index_col = 0)\n",
    "    print('Tax_df row num: ' + str(tax_df_orig.shape[0]))\n",
    "    \n",
    "    tax_df = order_tax(tax_df_orig)\n",
    "    tax_df = tax_df.apply(lambda row: get_cond(row), axis = 1)\n",
    "    tax_df['g'] = tax_df['g'].replace({0: 'Unknown'})\n",
    "    \n",
    "    ## Speciate df\n",
    "    speciate_df = pd.read_csv(path + spec_path, sep = '\\t', header = None)\n",
    "    print('Speciateit row num: ' + str(speciate_df.shape[0]))\n",
    "    speciate_df = order_speciate(speciate_df)\n",
    "    \n",
    "    ## Merge both dfs\n",
    "    merged_tax_df = tax_df.join(speciate_df)\n",
    "    \n",
    "    ## Create final name column\n",
    "    merged_tax_df['final_name'] = merged_tax_df.apply(add_final_name, axis = 1)\n",
    "    \n",
    "    ## Replace long names\n",
    "    long_names_lst = [name for name in merged_tax_df.final_name.unique() if type(name) == str and name.count('_') > 1]\n",
    "    print('Long names lst: ')\n",
    "    print(*long_names_lst, sep = '\\n')\n",
    "    \n",
    "    new_names = []\n",
    "    for name in long_names_lst:\n",
    "        new_name = get_new_name(name)\n",
    "        new_names.append(new_name)\n",
    "        \n",
    "    replacers = dict(zip(long_names_lst, new_names))\n",
    "    replacers.update({'Lactobacillus_ultunensis': 'Lactobacillus_crispatus'})\n",
    "\n",
    "    merged_tax_df['final_name'] = merged_tax_df['final_name'].replace(replacers)\n",
    "    merged_tax_df = merged_tax_df.loc[(merged_tax_df['p'] != 0)]\n",
    "    print('Merged tax df after removal of k_ ASVs: ' + str(merged_tax_df.shape[0]))\n",
    "    \n",
    "    long_names_lst = [name for name in merged_tax_df.final_name.unique() if name.count('_') > 1]\n",
    "    print((long_names_lst))\n",
    "    \n",
    "    return merged_tax_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a8e025c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spec_reads_table(path, merged_tax_df):\n",
    "    abun_df = pd.read_csv(path + asv_abun_file_name, index_col = 0)\n",
    "    print('Orig reads table columns: ' + str(abun_df.shape[1]))\n",
    "    \n",
    "    filt_abun_df = abun_df[merged_tax_df.index]\n",
    "    abun_df_t = filt_abun_df.T\n",
    "    spec_abun_df = pd.merge(merged_tax_df[['final_name']], abun_df_t, left_index = True, right_index = True)\n",
    "\n",
    "    spec_abun_df.reset_index(drop = True, inplace = True)\n",
    "    spec_abun_df.set_index('final_name', inplace = True)\n",
    "    spec_abun_df.index.name = None\n",
    "    spec_abun_df = spec_abun_df.T\n",
    "    print('Final reads table columns: ' + str(spec_abun_df.shape[1]))\n",
    "        \n",
    "    return spec_abun_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a0b5bb25",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_filtered(df):\n",
    "    filt_df = df.loc[df.sum(axis = 1) >= 1000]\n",
    "    filt_df02 = filt_df.loc[:, filt_df.sum(axis = 0) >= 100]\n",
    "    filt_df03 = filt_df02.drop(filt_df02.loc[filt_df02.sum(axis = 1) == 0].index, inplace = False)\n",
    "    \n",
    "    print('Shape after filter- rows: ' + str(filt_df03.shape[0]) + ', columns: ' + str(filt_df03.shape[1]))\n",
    "\n",
    "    return filt_df03"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "47d20fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_spec(df):\n",
    "    sum_spec_df = df.groupby(df.columns, axis = 1).sum()\n",
    "    print('Sum of species columns: ' + str(sum_spec_df.shape[1]))\n",
    "    \n",
    "    return sum_spec_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ae01a2b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_valencia_input(df):   \n",
    "    valencia_df = df.copy()\n",
    "    valencia_df.reset_index(inplace = True)\n",
    "    valencia_df = valencia_df.rename(columns = {'index':'sampleID'})\n",
    "    read_count_values = valencia_df.sum(axis = 1)\n",
    "    valencia_df.insert(loc = 1, column = 'read_count', value = read_count_values)\n",
    "    \n",
    "    return valencia_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0bd411c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_and_cmd_valencia(path, df):\n",
    "    valencia_file_name = 'valencia_df_' + date \n",
    "    df.to_csv(path + valencia_file_name + '.csv', index = False)\n",
    "    \n",
    "    in_file = path + valencia_file_name + '.csv'\n",
    "    out_file = path + valencia_file_name + '_res'\n",
    "    cmd = 'py C:/Users/morts/Documents/TAU/Python_Scripts/Manifold_Project/Valencia/Valencia.py -ref C:/Users/morts/Documents/TAU/Taxonomy/CST_centroids_19042023.csv -i ' + in_file + ' -o ' + out_file\n",
    "    print('*** cmd for VALENCIA ***')\n",
    "    print(cmd)\n",
    "    \n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "38b10779",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_func(name, path, spec_path):\n",
    "    ## Table with ASV and taxa final\n",
    "    merged_tax_df = get_spec_df(path, spec_path)\n",
    "    \n",
    "    ## Reads table with taxa names (after sum)\n",
    "    spec_abun_df = get_spec_reads_table(path, merged_tax_df)\n",
    "    \n",
    "    ## Filtering\n",
    "    filt_df = get_filtered(spec_abun_df)\n",
    "    sum_spec_df = sum_spec(filt_df)\n",
    "    \n",
    "    ## VALENCIA input\n",
    "    input_valencia = get_valencia_input(sum_spec_df)\n",
    "    save_and_cmd_valencia(path, input_valencia)\n",
    "    \n",
    "    print('****  ****  ****')\n",
    "    \n",
    "    return merged_tax_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fac23f88",
   "metadata": {},
   "source": [
    "#### Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "04ddba8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_lst = [db_path1, db_path2, db_path3, db_path4, db_path5, db_path6]\n",
    "speciate_lst = [speciate_filename1, speciate_filename2, speciate_filename3, speciate_filename4, speciate_filenam5, speciate_filename6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9623dd77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tax_df row num: 22633\n",
      "Speciateit row num: 22633\n",
      "Long names lst: \n",
      "Atopobium_sp_Atopobium_parvulum_Atopobium_rimae_Atopobium_vaginae\n",
      "g_Corynebacterium_1\n",
      "Lactobacillus_vaginalis_Lactobacillus_reuteri\n",
      "g_Rikenellaceae_RC9_gut_group\n",
      "g_Clostridium_sensu_stricto_1\n",
      "f_Bacteroidales_S24-7_group\n",
      "g_Ruminococcaceae_UCG-014\n",
      "f_Family_XI\n",
      "Lactobacillus_acidophilus_Lactobacillus_sp_Lactobacillus_taiwanensis_Lactobacillus_johnsonii\n",
      "f_Family_XIII\n",
      "g_Prevotella_7\n",
      "Lactobacillus_kitasatonis_Lactobacillus_gallinarum_Lactobacillus_crispatus\n",
      "g_Tyzzerella_4\n",
      "g_Ruminococcus_2\n",
      "g_Christensenellaceae_R-7_group\n",
      "f_Clostridiaceae_1\n",
      "g_Coprococcus_2\n",
      "g_Ruminiclostridium_5\n",
      "g_Ruminococcus_1\n",
      "g_Prevotella_6\n",
      "g_Ruminiclostridium_6\n",
      "g_Ruminococcaceae_NK4A214_group\n",
      "p_SR1_(Absconditabacteria)\n",
      "g_Prevotella_9\n",
      "f_Clostridiales_vadinBB60_group\n",
      "o_Mollicutes_RF9\n",
      "g_Treponema_2\n",
      "g_Ruminococcaceae_UCG-013\n",
      "g_Lachnoclostridium_5\n",
      "Lactobacillus_murinus_Lactobacillus_apodemi\n",
      "g_Family_XIII_AD3011_group\n",
      "g_Ruminococcaceae_UCG-010\n",
      "f_Clostridiaceae_3\n",
      "g_Candidatus_Saccharimonas\n",
      "g_Ruminiclostridium_1\n",
      "g_Blvii28_wastewater-sludge_group\n",
      "g_Coprococcus_1\n",
      "Lactobacillus_hamsteri_Lactobacillus_amylolyticus\n",
      "g_Lachnospiraceae_UCG-010\n",
      "g_Ruminiclostridium_9\n",
      "g_Selenomonas_3\n",
      "g_Butyrivibrio_2\n",
      "g_Candidatus_Stoquefichus\n",
      "Prevotella_genomosp_Prevotella_fusca_Prevotella_loescheii_Prevotella_conceptionensis\n",
      "g_Candidatus_Soleaferrea\n",
      "g_Ruminococcaceae_UCG-011\n",
      "o_Alphaproteobacteria_Incertae_Sedis\n",
      "g_Lachnospiraceae_NK3A20_group\n",
      "g_Defluviitaleaceae_UCG-011\n",
      "g_Candidatus_Captivus\n",
      "f_Bacteroidales_BS11_gut_group\n",
      "f_Clostridiaceae_2\n",
      "f_DA101_soil_group\n",
      "f_Aphagea_fa\n",
      "g_Candidatus_Protochlamydia\n",
      "g_Coprococcus_3\n",
      "Lactobacillus_fructivorans_Lactobacillus_acetotolerans\n",
      "c_Subgroup_6\n",
      "g_Tyzzerella_3\n",
      "g_Candidatus_Arthromitus\n",
      "g_dgA-11_gut_group\n",
      "Merged tax df after removal of k_ ASVs: 17495\n",
      "['f_Bacteroidales_S24-7_group', 'f_Family_XI', 'f_Family_XIII', 'f_Clostridiaceae_1', 'f_Clostridiales_vadinBB60_group', 'f_Clostridiaceae_3', 'f_Bacteroidales_BS11_gut_group', 'f_Clostridiaceae_2', 'f_DA101_soil_group', 'f_Aphagea_fa']\n",
      "Orig reads table columns: 22633\n",
      "Final reads table columns: 17495\n",
      "Shape after filter- rows: 3412, columns: 8638\n",
      "Sum of species columns: 278\n",
      "*** cmd for VALENCIA ***\n",
      "py C:/Users/morts/Documents/TAU/Python_Scripts/Manifold_Project/Valencia/Valencia.py -ref C:/Users/morts/Documents/TAU/Taxonomy/CST_centroids_19042023.csv -i C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Ravel_et_al.2013/Hiseq_sequencing/DADA2_results/valencia_df_16052023.csv -o C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Ravel_et_al.2013/Hiseq_sequencing/DADA2_results/valencia_df_16052023_res\n",
      "****  ****  ****\n",
      "***\n",
      "Tax_df row num: 6676\n",
      "Speciateit row num: 6676\n",
      "Long names lst: \n",
      "Atopobium_sp_Atopobium_parvulum_Atopobium_rimae_Atopobium_vaginae\n",
      "Lactobacillus_Cluster_1\n",
      "Lactobacillus_murinus_Lactobacillus_animalis\n",
      "Merged tax df after removal of k_ ASVs: 6545\n",
      "['Lactobacillus_Cluster_1']\n",
      "Orig reads table columns: 6676\n",
      "Final reads table columns: 6545\n",
      "Shape after filter- rows: 1573, columns: 1516\n",
      "Sum of species columns: 82\n",
      "*** cmd for VALENCIA ***\n",
      "py C:/Users/morts/Documents/TAU/Python_Scripts/Manifold_Project/Valencia/Valencia.py -ref C:/Users/morts/Documents/TAU/Taxonomy/CST_centroids_19042023.csv -i C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Ravel_et_al.2013/454_Sequencing/DADA2_results/DADA_output_files/valencia_df_16052023.csv -o C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Ravel_et_al.2013/454_Sequencing/DADA2_results/DADA_output_files/valencia_df_16052023_res\n",
      "****  ****  ****\n",
      "***\n",
      "Tax_df row num: 3598\n",
      "Speciateit row num: 3598\n",
      "Long names lst: \n",
      "Atopobium_sp_Atopobium_parvulum_Atopobium_rimae_Atopobium_vaginae\n",
      "Lactobacillus_vaginalis_Lactobacillus_reuteri\n",
      "Lactobacillus_murinus_Lactobacillus_apodemi\n",
      "Lactobacillus_acidophilus_Lactobacillus_sp_Lactobacillus_taiwanensis_Lactobacillus_johnsonii\n",
      "g_Rikenellaceae_RC9_gut_group\n",
      "f_Bacteroidales_S24-7_group\n",
      "Lactobacillus_kitasatonis_Lactobacillus_gallinarum_Lactobacillus_crispatus\n",
      "g_Corynebacterium_1\n",
      "g_Prevotella_7\n",
      "g_Candidatus_Captivus\n",
      "g_Prevotella_9\n",
      "Lactobacillus_zeae_Lactobacillus_casei_Lactobacillus_paracasei\n",
      "g_Ruminococcaceae_UCG-014\n",
      "g_Christensenellaceae_R-7_group\n",
      "g_Candidatus_Saccharimonas\n",
      "g_Clostridium_sensu_stricto_1\n",
      "f_Family_XIII\n",
      "g_Ruminiclostridium_6\n",
      "g_Ruminococcus_2\n",
      "g_Coprococcus_2\n",
      "g_Tyzzerella_4\n",
      "g_Ruminococcaceae_NK4A214_group\n",
      "g_Ruminococcus_1\n",
      "g_Ruminococcaceae_UCG-010\n",
      "g_Treponema_2\n",
      "g_Butyrivibrio_2\n",
      "f_Clostridiaceae_1\n",
      "g_Lachnospiraceae_UCG-010\n",
      "g_Prevotella_6\n",
      "g_Family_XIII_AD3011_group\n",
      "g_Selenomonas_4\n",
      "f_Clostridiales_vadinBB60_group\n",
      "f_Family_XI\n",
      "g_Clostridium_sensu_stricto_10\n",
      "g_Ruminococcaceae_UCG-013\n",
      "p_TM6_(Dependentiae)\n",
      "c_Subgroup_6\n",
      "f_NS11-12_marine_group\n",
      "o_Mollicutes_RF9\n",
      "g_Ruminiclostridium_9\n",
      "g_hgcI_clade\n",
      "g_Ruminiclostridium_5\n",
      "f_Unknown_Family\n",
      "Merged tax df after removal of k_ ASVs: 3544\n",
      "['f_Bacteroidales_S24-7_group', 'f_Family_XIII', 'f_Clostridiaceae_1', 'f_Clostridiales_vadinBB60_group', 'f_Family_XI', 'f_NS11-12_marine_group', 'f_Unknown_Family']\n",
      "Orig reads table columns: 3598\n",
      "Final reads table columns: 3544\n",
      "Shape after filter- rows: 706, columns: 1576\n",
      "Sum of species columns: 137\n",
      "*** cmd for VALENCIA ***\n",
      "py C:/Users/morts/Documents/TAU/Python_Scripts/Manifold_Project/Valencia/Valencia.py -ref C:/Users/morts/Documents/TAU/Taxonomy/CST_centroids_19042023.csv -i C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Carter_et_al.2022/DADA2_results/valencia_df_16052023.csv -o C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Carter_et_al.2022/DADA2_results/valencia_df_16052023_res\n",
      "****  ****  ****\n",
      "***\n",
      "Tax_df row num: 4851\n",
      "Speciateit row num: 4851\n",
      "Long names lst: \n",
      "Lactobacillus_Cluster_1\n",
      "Atopobium_sp_Atopobium_parvulum_Atopobium_rimae_Atopobium_vaginae\n",
      "Lactobacillus_gallinarum_Lactobacillus_helveticus\n",
      "g_Prevotella_7\n",
      "Lactobacillus_salivarius_Lactobacillus_phage\n",
      "Merged tax df after removal of k_ ASVs: 4840\n",
      "['Lactobacillus_Cluster_1']\n",
      "Orig reads table columns: 4851\n",
      "Final reads table columns: 4840\n",
      "Shape after filter- rows: 385, columns: 875\n",
      "Sum of species columns: 54\n",
      "*** cmd for VALENCIA ***\n",
      "py C:/Users/morts/Documents/TAU/Python_Scripts/Manifold_Project/Valencia/Valencia.py -ref C:/Users/morts/Documents/TAU/Taxonomy/CST_centroids_19042023.csv -i C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Ravel_et_al.2011/DADA2_results/valencia_df_16052023.csv -o C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Ravel_et_al.2011/DADA2_results/valencia_df_16052023_res\n",
      "****  ****  ****\n",
      "***\n"
     ]
    }
   ],
   "source": [
    "for i, path in enumerate(db_lst):\n",
    "    df = all_func('test', db_lst[i], speciate_lst[i])\n",
    "    print('***')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc9601e2",
   "metadata": {},
   "source": [
    "#### Check results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "376d335c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tax_df row num: 37719\n",
      "Speciateit row num: 37719\n",
      "Long names lst: \n",
      "Atopobium_sp_Atopobium_parvulum_Atopobium_rimae_Atopobium_vaginae\n",
      "Lactobacillus_Cluster_1\n",
      "Lactobacillus_gallinarum_Lactobacillus_helveticus\n",
      "Lactobacillus_zeae_Lactobacillus_casei_Lactobacillus_paracasei\n",
      "Lactobacillus_salivarius_Lactobacillus_phage\n",
      "g_Prevotella_7\n",
      "Lactobacillus_murinus_Lactobacillus_animalis\n",
      "Prevotella_loescheii_Prevotella_genomosp_Prevotella_conceptionensis\n",
      "Lactobacillus_parabuchneri_Lactobacillus_rapi\n",
      "Merged tax df after removal of k_ ASVs: 37240\n",
      "['Lactobacillus_Cluster_1']\n",
      "Orig reads table columns: 37719\n",
      "Final reads table columns: 37240\n",
      "Shape after filter- rows: 3879, columns: 7835\n",
      "Sum of species columns: 278\n",
      "*** cmd for VALENCIA ***\n",
      "py C:/Users/morts/Documents/TAU/Python_Scripts/Manifold_Project/Valencia/Valencia.py -ref C:/Users/morts/Documents/TAU/Taxonomy/CST_centroids_19042023.csv -i C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Serrano_et_al.2019/noDuplicatedID/DADA2_results/valencia_df_16052023.csv -o C:/Users/morts/Documents/TAU/Manifold_Project/Data/SPECIFIC_PAPERS/Serrano_et_al.2019/noDuplicatedID/DADA2_results/valencia_df_16052023_res\n",
      "****  ****  ****\n"
     ]
    }
   ],
   "source": [
    "df = all_func('test', db_path5, speciate_filename5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "74fa6492",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "      <th>cond_spec</th>\n",
       "      <th>speciate_spec</th>\n",
       "      <th>final_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASV82</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Gordonibacter_sp_Gordonibacter_pamelaeae</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV277</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Gordonibacter_sp_Gordonibacter_pamelaeae</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV473</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Parvibacter_caecicola</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV867</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Gordonibacter_sp_Gordonibacter_pamelaeae</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV1082</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Gordonibacter_sp_Gordonibacter_pamelaeae</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV36941</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Cellulosibacter_sp</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37484</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Limnochorda_pilosa</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37515</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Oscillospirales</td>\n",
       "      <td>Oscillospiraceae</td>\n",
       "      <td>UCG-002</td>\n",
       "      <td>0</td>\n",
       "      <td>g_UCG-002</td>\n",
       "      <td>Oscillibacter_sp</td>\n",
       "      <td>f_Oscillospiraceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37531</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Coriobacteriia</td>\n",
       "      <td>Coriobacteriales</td>\n",
       "      <td>Eggerthellaceae</td>\n",
       "      <td>DNF00809</td>\n",
       "      <td>0</td>\n",
       "      <td>g_DNF00809</td>\n",
       "      <td>Gorbachella_massiliensis</td>\n",
       "      <td>f_Eggerthellaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37597</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Peptostreptococcales-Tissierellales</td>\n",
       "      <td>Anaerovoracaceae</td>\n",
       "      <td>Family XIII UCG-001</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Family XIII UCG-001</td>\n",
       "      <td>Eubacterium_sp</td>\n",
       "      <td>f_Anaerovoracaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>375 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k                 p               c  \\\n",
       "ASV82     Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV277    Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV473    Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV867    Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV1082   Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "...            ...               ...             ...   \n",
       "ASV36941  Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV37484  Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV37515  Bacteria        Firmicutes      Clostridia   \n",
       "ASV37531  Bacteria  Actinobacteriota  Coriobacteriia   \n",
       "ASV37597  Bacteria        Firmicutes      Clostridia   \n",
       "\n",
       "                                            o                 f  \\\n",
       "ASV82                        Coriobacteriales   Eggerthellaceae   \n",
       "ASV277                       Coriobacteriales   Eggerthellaceae   \n",
       "ASV473                       Coriobacteriales   Eggerthellaceae   \n",
       "ASV867                       Coriobacteriales   Eggerthellaceae   \n",
       "ASV1082                      Coriobacteriales   Eggerthellaceae   \n",
       "...                                       ...               ...   \n",
       "ASV36941                     Coriobacteriales   Eggerthellaceae   \n",
       "ASV37484                     Coriobacteriales   Eggerthellaceae   \n",
       "ASV37515                      Oscillospirales  Oscillospiraceae   \n",
       "ASV37531                     Coriobacteriales   Eggerthellaceae   \n",
       "ASV37597  Peptostreptococcales-Tissierellales  Anaerovoracaceae   \n",
       "\n",
       "                            g  s              cond_spec  \\\n",
       "ASV82                DNF00809  0             g_DNF00809   \n",
       "ASV277               DNF00809  0             g_DNF00809   \n",
       "ASV473               DNF00809  0             g_DNF00809   \n",
       "ASV867               DNF00809  0             g_DNF00809   \n",
       "ASV1082              DNF00809  0             g_DNF00809   \n",
       "...                       ... ..                    ...   \n",
       "ASV36941             DNF00809  0             g_DNF00809   \n",
       "ASV37484             DNF00809  0             g_DNF00809   \n",
       "ASV37515              UCG-002  0              g_UCG-002   \n",
       "ASV37531             DNF00809  0             g_DNF00809   \n",
       "ASV37597  Family XIII UCG-001  0  g_Family XIII UCG-001   \n",
       "\n",
       "                                     speciate_spec          final_name  \n",
       "ASV82     Gordonibacter_sp_Gordonibacter_pamelaeae   f_Eggerthellaceae  \n",
       "ASV277    Gordonibacter_sp_Gordonibacter_pamelaeae   f_Eggerthellaceae  \n",
       "ASV473                       Parvibacter_caecicola   f_Eggerthellaceae  \n",
       "ASV867    Gordonibacter_sp_Gordonibacter_pamelaeae   f_Eggerthellaceae  \n",
       "ASV1082   Gordonibacter_sp_Gordonibacter_pamelaeae   f_Eggerthellaceae  \n",
       "...                                            ...                 ...  \n",
       "ASV36941                        Cellulosibacter_sp   f_Eggerthellaceae  \n",
       "ASV37484                        Limnochorda_pilosa   f_Eggerthellaceae  \n",
       "ASV37515                          Oscillibacter_sp  f_Oscillospiraceae  \n",
       "ASV37531                  Gorbachella_massiliensis   f_Eggerthellaceae  \n",
       "ASV37597                            Eubacterium_sp  f_Anaerovoracaceae  \n",
       "\n",
       "[375 rows x 10 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['cond_spec'].str.contains('00')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8a9e55f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "      <th>cond_spec</th>\n",
       "      <th>speciate_spec</th>\n",
       "      <th>final_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASV4</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV24</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV34</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV39</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37591</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37598</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37630</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37691</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Veillonellales-Selenomonadales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>f_Veillonellaceae</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV37712</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Shuttleworthia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Shuttleworthia</td>\n",
       "      <td>BVAB1</td>\n",
       "      <td>BVAB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1418 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 k           p              c                               o  \\\n",
       "ASV4      Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV6      Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV24     Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV34     Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV39     Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "...            ...         ...            ...                             ...   \n",
       "ASV37591  Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV37598  Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV37630  Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "ASV37691  Bacteria  Firmicutes  Negativicutes  Veillonellales-Selenomonadales   \n",
       "ASV37712  Bacteria  Firmicutes     Clostridia                  Lachnospirales   \n",
       "\n",
       "                        f               g  s          cond_spec speciate_spec  \\\n",
       "ASV4      Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV6      Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV24     Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV34     Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV39     Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "...                   ...             ... ..                ...           ...   \n",
       "ASV37591  Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV37598  Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV37630  Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "ASV37691  Veillonellaceae         Unknown  0  f_Veillonellaceae         BVAB1   \n",
       "ASV37712  Lachnospiraceae  Shuttleworthia  0   g_Shuttleworthia         BVAB1   \n",
       "\n",
       "         final_name  \n",
       "ASV4          BVAB1  \n",
       "ASV6          BVAB1  \n",
       "ASV24         BVAB1  \n",
       "ASV34         BVAB1  \n",
       "ASV39         BVAB1  \n",
       "...             ...  \n",
       "ASV37591      BVAB1  \n",
       "ASV37598      BVAB1  \n",
       "ASV37630      BVAB1  \n",
       "ASV37691      BVAB1  \n",
       "ASV37712      BVAB1  \n",
       "\n",
       "[1418 rows x 10 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['speciate_spec'].str.contains('BVAB')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "130d81d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "test1 = all_func('9', db_path1, speciate_filename1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1e6b4e56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tax_df row num: 6676\n",
      "Speciateit row num: 6676\n",
      "Long names lst: \n",
      "g_Prevotella_7\n",
      "Lactobacillus_Cluster_1\n",
      "Lactobacillus_murinus_Lactobacillus_animalis\n",
      "Merged tax df after removal of k_ ASVs: 6545\n",
      "['Lactobacillus_Cluster_1']\n"
     ]
    }
   ],
   "source": [
    "test2 = get_spec_df(db_path1, speciate_filename1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ed3d506a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "      <th>cond_spec</th>\n",
       "      <th>speciate_spec</th>\n",
       "      <th>final_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASV1179</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV1526</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Thermicanus_aegyptius</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV1575</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV1623</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV1734</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6486</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6500</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Salinivirga_cyanobacteriivorans</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6537</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6637</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Segniliparus_rotundus</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6669</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "      <td>Pasteuria_goettingianae</td>\n",
       "      <td>g_Gardnerella</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                k                 p               c                  o  \\\n",
       "ASV1179  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV1526  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV1575  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV1623  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV1734  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "...           ...               ...             ...                ...   \n",
       "ASV6486  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV6500  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV6537  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV6637  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV6669  Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "\n",
       "                          f            g  s      cond_spec  \\\n",
       "ASV1179  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV1526  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV1575  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV1623  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV1734  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "...                     ...          ... ..            ...   \n",
       "ASV6486  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV6500  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV6537  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV6637  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "ASV6669  Bifidobacteriaceae  Gardnerella  0  g_Gardnerella   \n",
       "\n",
       "                           speciate_spec     final_name  \n",
       "ASV1179          Pasteuria_goettingianae  g_Gardnerella  \n",
       "ASV1526            Thermicanus_aegyptius  g_Gardnerella  \n",
       "ASV1575          Pasteuria_goettingianae  g_Gardnerella  \n",
       "ASV1623          Pasteuria_goettingianae  g_Gardnerella  \n",
       "ASV1734          Pasteuria_goettingianae  g_Gardnerella  \n",
       "...                                  ...            ...  \n",
       "ASV6486          Pasteuria_goettingianae  g_Gardnerella  \n",
       "ASV6500  Salinivirga_cyanobacteriivorans  g_Gardnerella  \n",
       "ASV6537          Pasteuria_goettingianae  g_Gardnerella  \n",
       "ASV6637            Segniliparus_rotundus  g_Gardnerella  \n",
       "ASV6669          Pasteuria_goettingianae  g_Gardnerella  \n",
       "\n",
       "[117 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2.loc[test2['final_name'] == 'g_Gardnerella']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "16290175",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "      <th>cond_spec</th>\n",
       "      <th>speciate_spec</th>\n",
       "      <th>final_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASV1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV2</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV3</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>vaginalis</td>\n",
       "      <td>Gardnerella_vaginalis</td>\n",
       "      <td>Gardnerella_vaginalis</td>\n",
       "      <td>Gardnerella_vaginalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV4</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "      <td>Lactobacillus_iners</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV5</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacteriales</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Gardnerella</td>\n",
       "      <td>vaginalis</td>\n",
       "      <td>Gardnerella_vaginalis</td>\n",
       "      <td>Gardnerella_vaginalis</td>\n",
       "      <td>Gardnerella_vaginalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6671</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Bacteroidaceae</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Bacteroides</td>\n",
       "      <td>Bacteroides_sp_Bacteroides_vulgatus</td>\n",
       "      <td>g_Bacteroides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6672</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Blautia</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Blautia</td>\n",
       "      <td>Blautia_producta_Blautia_hansenii_Blautia_sp_B...</td>\n",
       "      <td>g_Blautia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6673</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Corynebacterium</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Corynebacterium</td>\n",
       "      <td>Corynebacterium_genitalium</td>\n",
       "      <td>g_Corynebacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6674</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Mobiluncus</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Mobiluncus</td>\n",
       "      <td>Mobiluncus_curtisii</td>\n",
       "      <td>g_Mobiluncus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV6675</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Mobiluncus</td>\n",
       "      <td>0</td>\n",
       "      <td>g_Mobiluncus</td>\n",
       "      <td>Mobiluncus_curtisii</td>\n",
       "      <td>g_Mobiluncus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6545 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                k                 p               c                  o  \\\n",
       "ASV1     Bacteria        Firmicutes         Bacilli    Lactobacillales   \n",
       "ASV2     Bacteria        Firmicutes         Bacilli    Lactobacillales   \n",
       "ASV3     Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "ASV4     Bacteria        Firmicutes         Bacilli    Lactobacillales   \n",
       "ASV5     Bacteria  Actinobacteriota  Actinobacteria  Bifidobacteriales   \n",
       "...           ...               ...             ...                ...   \n",
       "ASV6671  Bacteria      Bacteroidota     Bacteroidia      Bacteroidales   \n",
       "ASV6672  Bacteria        Firmicutes      Clostridia     Lachnospirales   \n",
       "ASV6673  Bacteria  Actinobacteriota  Actinobacteria  Corynebacteriales   \n",
       "ASV6674  Bacteria  Actinobacteriota  Actinobacteria    Actinomycetales   \n",
       "ASV6675  Bacteria  Actinobacteriota  Actinobacteria    Actinomycetales   \n",
       "\n",
       "                          f                g          s  \\\n",
       "ASV1       Lactobacillaceae    Lactobacillus      iners   \n",
       "ASV2       Lactobacillaceae    Lactobacillus      iners   \n",
       "ASV3     Bifidobacteriaceae      Gardnerella  vaginalis   \n",
       "ASV4       Lactobacillaceae    Lactobacillus      iners   \n",
       "ASV5     Bifidobacteriaceae      Gardnerella  vaginalis   \n",
       "...                     ...              ...        ...   \n",
       "ASV6671      Bacteroidaceae      Bacteroides          0   \n",
       "ASV6672     Lachnospiraceae          Blautia          0   \n",
       "ASV6673  Corynebacteriaceae  Corynebacterium          0   \n",
       "ASV6674    Actinomycetaceae       Mobiluncus          0   \n",
       "ASV6675    Actinomycetaceae       Mobiluncus          0   \n",
       "\n",
       "                     cond_spec  \\\n",
       "ASV1       Lactobacillus_iners   \n",
       "ASV2       Lactobacillus_iners   \n",
       "ASV3     Gardnerella_vaginalis   \n",
       "ASV4       Lactobacillus_iners   \n",
       "ASV5     Gardnerella_vaginalis   \n",
       "...                        ...   \n",
       "ASV6671          g_Bacteroides   \n",
       "ASV6672              g_Blautia   \n",
       "ASV6673      g_Corynebacterium   \n",
       "ASV6674           g_Mobiluncus   \n",
       "ASV6675           g_Mobiluncus   \n",
       "\n",
       "                                             speciate_spec  \\\n",
       "ASV1                                   Lactobacillus_iners   \n",
       "ASV2                                   Lactobacillus_iners   \n",
       "ASV3                                 Gardnerella_vaginalis   \n",
       "ASV4                                   Lactobacillus_iners   \n",
       "ASV5                                 Gardnerella_vaginalis   \n",
       "...                                                    ...   \n",
       "ASV6671                Bacteroides_sp_Bacteroides_vulgatus   \n",
       "ASV6672  Blautia_producta_Blautia_hansenii_Blautia_sp_B...   \n",
       "ASV6673                         Corynebacterium_genitalium   \n",
       "ASV6674                                Mobiluncus_curtisii   \n",
       "ASV6675                                Mobiluncus_curtisii   \n",
       "\n",
       "                    final_name  \n",
       "ASV1       Lactobacillus_iners  \n",
       "ASV2       Lactobacillus_iners  \n",
       "ASV3     Gardnerella_vaginalis  \n",
       "ASV4       Lactobacillus_iners  \n",
       "ASV5     Gardnerella_vaginalis  \n",
       "...                        ...  \n",
       "ASV6671          g_Bacteroides  \n",
       "ASV6672              g_Blautia  \n",
       "ASV6673      g_Corynebacterium  \n",
       "ASV6674           g_Mobiluncus  \n",
       "ASV6675           g_Mobiluncus  \n",
       "\n",
       "[6545 rows x 10 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9d198d4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spec = 'Eggerthe'\n",
    "'f_Eggerthellaceae'\n",
    "[col for col in df.columns if spec in col]\n",
    "# any(test.apply(lambda row: row.astype(str).str.contains(spec).any(), axis=1))\n",
    "# test[test['cond_spec'].str.contains(spec)]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
